src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,494 @@
+}
+
+void DockerVirtualMachinePrivate::doAddPortForwarding(const QString &ruleName, const QString &protocol,
+        quint16 hostPort, quint16 emulatorVmPort, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(protocol)
+    Q_UNUSED(hostPort)
+    Q_UNUSED(emulatorVmPort)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRemovePortForwarding(const QString &ruleName,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortForwarding(VirtualMachinePrivate::ReservedPort which,
+        quint16 port, const QObject *context, const Functor<bool> &functor)
+{
+    QPair<quint16, quint16>& mapping = m_reservedPortMappings[which];
+    mapping.first = port;
+
+    buildWithLabel(label(which), QString::number(port), context, functor);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortListForwarding(
+        VirtualMachinePrivate::ReservedPortList which, const QList<Utils::Port> &ports,
+        const QObject *context, const Functor<const QMap<QString, quint16> &, bool> &functor)
+{
+    Q_UNUSED(which)
+    Q_UNUSED(ports)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRestoreSnapshot(const QString &snapshotName, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_UNUSED(snapshotName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::SharedPath which)
+{
+    QString label;
+    switch (which) {
+    case SharedSrc:
+        label = Constants::BUILD_ENGINE_SHARED_SRC;
+        break;
+    case SharedHome:
+        label = Constants::BUILD_ENGINE_SHARED_HOME;
+        break;
+    case SharedConfig:
+        label = Constants::BUILD_ENGINE_SHARED_CONFIG;
+        break;
+    case SharedSsh:
+        label = Constants::BUILD_ENGINE_SHARED_SSH;
+        break;
+    case SharedTargets:
+        label = Constants::BUILD_ENGINE_SHARED_TARGET;
+    break;
+    }
+    return label;
+}
+
+QString DockerVirtualMachinePrivate::label(VirtualMachinePrivate::ReservedPort which)
+{
+    QString label;
+    switch (which) {
+    case SshPort:
+        label = Constants::BUILD_ENGINE_SSH_PORT;
+        break;
+    case WwwPort:
+        label = Constants::BUILD_ENGINE_WWW_PORT;
+        break;
+    }
+    return label;
+}
+
+void DockerVirtualMachinePrivate::buildWithLabel(const QString& key, const QString& value,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_Q(DockerVirtualMachine);
+    QStringList arguments;
+    arguments.append(COMMAND_BUILD);
+    arguments.append(DOCKERFILE_PREFIX + Sdk::installationPath() + DOCKERFILE_POSTFIX);
```

Comment:
It seems possible to pass "-" instead and write the dockerfile on stdin with something like

```
const QByteArray dockerfile = ...;
connect(runner.get(), &QProcess::started, [=]() {
    runner->process()->write(dockerfile);
    runner->process()->closeWriteChannel();
});
```

Reply:
Implemented now this way